skorch/tests/test_net.py

```diff
@@ -2620,33 +2746,597 @@ def initialize_module(self, *args, **kwargs):
+
+            def initialize_criterion(self):
+                super().initialize_criterion()
+                params = self.get_params_for('mycriterion')
+                self.mycriterion_ = self.mycriterion(**params)
+                return self
+
+        class MyModule(nn.Module):
+            """Custom module to test device even without GPU"""
+            def __init__(self, x=1):
+                super().__init__()
+                self.lin = nn.Linear(x, 1)  # modules need parameters
+                self.device = 'cpu'
+
+            def to(self, device):
+                self.device = device
+                return self
+
+        # first normal CPU
+        net = MyNet(
+            module=MyModule,
+            mymodule=MyModule,
+            criterion=MyModule,
+            mycriterion=MyModule,
+        ).initialize()
+        assert net.mymodule_.device == 'cpu'
+        assert net.mycriterion_.device == 'cpu'
+
+        # now try other device
+        net = MyNet(
+            module=MyModule,
+            mymodule=MyModule,
+            device='foo',
+            criterion=MyModule,
+            mycriterion=MyModule,
+        ).initialize()
+        assert net.mymodule_.device == 'foo'
+        assert net.mycriterion_.device == 'foo'
+
+        net.set_params(mymodule__x=3)
+        assert net.mymodule_.device == 'foo'
+        assert net.mycriterion_.device == 'foo'
+
+    def test_custom_modules_and_criteria_training_mode_set_correctly(
+            self, net_cls, module_cls, data,
+    ):
+        # custom modules and criteria should be set to training/eval mode
+        # correctly depending on the stage of training/validation/inference
+        from skorch.callbacks import Callback
+
+        class MyNet(net_cls):
+            """Net with custom mymodule and mycriterion"""
+            def initialize_module(self):
+                super().initialize_module()
+                self.mymodule_ = module_cls()
+                return self
+
+            def initialize_criterion(self):
+                super().initialize_criterion()
+                self.mycriterion_ = module_cls()
+                return self
+
+            def evaluation_step(self, batch, training=False):
+                y_pred = super().evaluation_step(batch, training=training)
+                assert_net_training_mode(self, training=training)
+                return y_pred
+
+        def assert_net_training_mode(net, training=True):
+            if training:
+                check = lambda module: module.training is True
+            else:
+                check = lambda module: module.training is False
+            assert check(net.module_)
+            assert check(net.mymodule_)
+            assert check(net.criterion_)
+            assert check(net.mycriterion_)
+
+        class CheckTrainingCallback(Callback):
+            def on_batch_end(self, net, batch, training, **kwargs):
+                assert_net_training_mode(net, training=training)
```

Comment:
I guess the easiest way to assert that `on_batch_end` is actually being called is to use the `nonlocal` trick from above.

Reply:
Hmm, we test more than just if the method is called. Therefore, I don't see how `nonlocal` would help to simplify the test here. Maybe it's just too late in the evening for me :)

Anyway, I could simplify the test by removing the callback completely and instead override `on_batch_end` of the net itself.